options/options_helper.cc

```diff
@@ -1015,28 +916,26 @@ Status StringToMap(const std::string& opts_str,
 }
 
 Status GetStringFromStruct(
-    const ConfigOptions& cfg_options, const void* const opt_ptr,
+    const ConfigOptions& config_options, const void* const opt_ptr,
     const std::unordered_map<std::string, OptionTypeInfo>& type_info,
     std::string* opt_string) {
   assert(opt_string);
   opt_string->clear();
   for (const auto iter : type_info) {
     const auto& opt_info = iter.second;
-    if (opt_info.IsDeprecated()) {
-      // If the option is no longer used in rocksdb and marked as deprecated,
-      // we skip it in the serialization.
-      continue;
-    }
-    const char* opt_address =
-        reinterpret_cast<const char*>(opt_ptr) + opt_info.offset;
-    std::string value;
-    bool result =
-        SerializeSingleOptionHelper(opt_address, opt_info.type, &value);
-    if (result) {
-      opt_string->append(iter.first + "=" + value + cfg_options.delimiter);
-    } else {
-      return Status::InvalidArgument("failed to serialize %s\n",
-                                     iter.first.c_str());
+    // If the option is no longer used in rocksdb and marked as deprecated,
+    // we skip it in the serialization.
+    if (opt_info.ShouldSerialize()) {
+      const char* opt_addr =
+          reinterpret_cast<const char*>(opt_ptr) + opt_info.offset;
+      std::string value;
+      Status s = opt_info.SerializeOption(config_options, iter.first, opt_addr,
+                                          &value);
+      if (s.ok()) {
+        opt_string->append(iter.first + "=" + value + config_options.delimiter);
+      } else {
+        return s;
```

Comment:
Should we return return `Status::InvalidArgument("failed to serialize %s\n", iter.first.c_str())` here?

Reply:
If we do that, we lose information.  The OptionTypeInfo::SerializeOption returned a specific status.  If we wrap it with a generic status, we have no idea what the lower level status returned.